### PR TITLE
Feature/intygfv 14791 Uncertain date component

### PR DESCRIPTION
**Pull request by SopraErik**
Might have missed something in the general flow,or structure of things? 
Styling of text input for date has a slightly diferent height as compared to the dopdowns for year and month. This might lie in the css framework?
Elements placed in grid, since flex layout would render dropdowns invisible, for some reason prabably also due to css framework.

### CODE REVIEW

File: packages/webcert/src/feature/certificate/Inputs/UeUncertainDate.tsx [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+  const dispatch = useAppDispatch()
+  const [selectedYear, setSelectedYear] = React.useState(getYear((question.value as ValueDate).date))
+  const [selectedMonth, setSelectedMonth] = React.useState(getMonth((question.value as ValueDate).date))
+  const [disabledMonth, setDisabledMonth] = React.useState(
+    disabled || getYear((question.value as ValueDate).date) === '0000' || getYear((question.value as ValueDate).date) === ''
+  )
+
+  const years: ConfigUeDropdownItem[] = [{ id: '', label: 'Ange år' }]
+
+  if (config.unknownYear) years.push({ id: '0000', label: '0000 (ej känt)' })
```
Comment (HelenaEdblad) [inline packages/webcert/src/feature/certificate/Inputs/UeUncertainDate.tsx:33]: I'm so boring now, I know ;) But I think it's hard to read if-statements like these. They look cool, but I don't think we write if-statements like that anywhere else in the code. What do you think?

File: packages/webcert/src/feature/certificate/Inputs/UeUncertainDate.tsx [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    const strMonth = '0' + i.toString()
+    const month = strMonth.substring(strMonth.length - 2)
+    months.push({ id: month, label: month })
+  }
+
+  const handleYearChange = (value: string) => {
+    setSelectedYear(value)
+    if (value === '') setSelectedMonth('')
+    if (value === '0000') setSelectedMonth('00')
+    setDisabledMonth(disabled || value === '0000' || value === '')
```
Comment (HelenaEdblad) [inline packages/webcert/src/feature/certificate/Inputs/UeUncertainDate.tsx:50]: Probably just me being confused, but which one is month set to? Disabled, 0000 or ''?

File: packages/webcert/src/feature/certificate/Inputs/UeUncertainDate.tsx [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+  )
+}
+
+const getYear = (date: string) => {
+  let year = ''
+  if (date && date.indexOf('-') === 4) year = date.split('-')[0]
+  return year
+}
+const getMonth = (date: string) => {
+  if (date && date.indexOf('-') === 4 && getYear(date) !== '0000' && getYear(date) !== '') {
```
Comment (HelenaEdblad) [inline packages/webcert/src/feature/certificate/Inputs/UeUncertainDate.tsx:108]: Is it possible to make what's inside the parentheses into a variable or somehow make it easier to read? :)

**Review by HelenaEdblad — CHANGES_REQUESTED** [review]
Nice! Apart from my comments in the code we are also missing tests. 
And, I saw that there is an error in the Storybook when you click on Dödsbevis. 
db

File: packages/webcert/src/feature/certificate/Inputs/UeUncertainDate.tsx (hunk 2 above)
Comment (SopraErik) [reply packages/webcert/src/feature/certificate/Inputs/UeUncertainDate.tsx:50]: month is set to not selected ('', ange månad) if year is not selected ('', ange år)
month is set to unknown ('00') if year is set to unknown ('0000')
the month control is disabled if either the whole control is disabled, or if year is not selected ('') or if year is unknown ('0000')

File: packages/webcert/src/feature/certificate/Inputs/UeUncertainDate.tsx (hunk 3 above)
Comment (SopraErik) [reply packages/webcert/src/feature/certificate/Inputs/UeUncertainDate.tsx]: I split it upp and use a regEx instead now, I think its better and more readable

File: packages/webcert/src/feature/certificate/Inputs/UeUncertainDate.tsx (hunk 1 above)
Comment (SopraErik) [reply packages/webcert/src/feature/certificate/Inputs/UeUncertainDate.tsx]: Just old habit. hopefully more readable now

File: packages/webcert/src/feature/certificate/Inputs/UeUncertainDate.tsx [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+  hasValidationError?: boolean
+  disabled?: boolean
+  question: CertificateDataElement
+}
+
+const UeUncertainDate: React.FC<Props> = (props) => {
+  const { question, hasValidationError, disabled } = props
+  const config = question.config as ConfigureUeUncertainDate
+  const dispatch = useAppDispatch()
+  const [selectedYear, setSelectedYear] = React.useState(getYear((question.value as ValueDate).date))
```
[the remark this answers is not in the document]
Comment (SopraErik) [reply packages/webcert/src/feature/certificate/Inputs/UeUncertainDate.tsx]: Of course, fixed now

File: packages/webcert/src/feature/certificate/Inputs/UeUncertainDate.tsx [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+const ValidationWrapper = styled.div`
+  flex: 0 !important;
+  flex-basis: 100% !important;
+  padding-bottom: 16px;
+  margin-top: 0;
+`
+
+export interface Props {
+  hasValidationError?: boolean
```
[the remark this answers is not in the document]
Comment (SopraErik) [reply packages/webcert/src/feature/certificate/Inputs/UeUncertainDate.tsx]: fixed according to your suggestion

File: packages/webcert/src/feature/certificate/Inputs/UeUncertainDate.tsx [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+  disabled?: boolean
+  question: CertificateDataElement
+}
+
+const UeUncertainDate: React.FC<Props> = (props) => {
+  const { question, hasValidationError, disabled } = props
```
[the remark this answers is not in the document]
Comment (SopraErik) [reply packages/webcert/src/feature/certificate/Inputs/UeUncertainDate.tsx]: fixed according to your suggestion

File: packages/webcert/src/feature/certificate/Inputs/UeUncertainDate.tsx [hunk 7]
```diff
@@ -0,0 +1,127 @@
+import { CertificateDataElement, Dropdown, ConfigureUeUncertainDate, QuestionValidationTexts, ValueDate } from '@frontend/common'
+import { ConfigUeDropdownItem } from '@frontend/common/src/types/certificate'
+import React from 'react'
+import styled from 'styled-components/macro'
+import { updateCertificateDataElement } from '../../../store/certificate/certificateActions'
+import { useAppDispatch } from '../../../store/store'
```
[the remark this answers is not in the document]
Comment (SopraErik) [reply packages/webcert/src/feature/certificate/Inputs/UeUncertainDate.tsx:8]: useAppDispatch is a typed wrapper fror useDispatch, hence from a typescript perspective it is probably better to use useAppDispatch to ensure typing is ok

[the remark this answers is not in the document]
Comment (SamuelTilly) [reply packages/webcert/src/feature/certificate/Inputs/UeUncertainDate.tsx:8]: This is interesting, looks like this is the recommended way of doing things when using `useDispatch` in a typescript project. https://react-redux.js.org/using-react-redux/usage-with-typescript#define-typed-hooks

We should probably start doing it more this way, and probably even look at wrapping `useSelector` in a `useAppSelector`

**Review by HelenaEdblad — APPROVED** [review]

**Pull request merged**